File: backend/NodeAPI/src/domain/entities/StudySession.ts
import { InvalidMinutesStudiedError,InvalidTasksCompletedError } from "../errors/StudySessionError";

export class StudySession {

    private studySessionId: number | undefined;
    private dateSession: Date;
    private minutesStudied: number;
    private tasksCompleted: number;
    private userId: number;

    private constructor(
        minutesStudied: number,
        tasksCompleted: number,
        userId: number,
        dateSession?: Date,
        studySessionId?: number
    ) {
        this.minutesStudied = minutesStudied;
        this.tasksCompleted = tasksCompleted;
        this.userId = userId;
        this.dateSession = dateSession ?? new Date();
        this.studySessionId = studySessionId;
    }

    static create(minutesStudied: number, tasksCompleted: number, userId: number): StudySession {
        if (minutesStudied <= 0) throw new InvalidMinutesStudiedError();
        if (tasksCompleted < 0) throw new InvalidTasksCompletedError();
        return new StudySession(minutesStudied, tasksCompleted, userId);
    }

    static restore(
        studySessionId: number,
        minutesStudied: number,
        tasksCompleted: number,
        userId: number,
        dateSession: Date
    ): StudySession {
        return new StudySession(minutesStudied, tasksCompleted, userId, dateSession, studySessionId);
    }

    getStudySessionId(): number | undefined { return this.studySessionId; }
    getDateSession(): Date { return this.dateSession; }
    getMinutesStudied(): number { return this.minutesStudied; }
    getTasksCompleted(): number { return this.tasksCompleted; }
    getUserId(): number { return this.userId; }

}

File: backend/JavaAPI/javaapi/src/main/java/com/devcoachai/javaapi/domain/entities/StudySession.java
package com.devcoachai.javaapi.domain.entities;

import jakarta.persistence.*;

import java.time.LocalDateTime;

@Entity
@Table(name = "\"StudySession\"")
public class StudySession {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id")
    private Integer id;
    @Column(name = "date")
    private LocalDateTime dateSession;
    @Column(name = "minutesStudied")
    private Integer minutesStudied;
    @Column(name = "tasksCompleted")
    private Integer tasksCompleted;
    @Column(name = "userId")
    private Integer userId;

    public StudySession(){}

    public Integer getId() {return id;}
    public LocalDateTime getDateSession() {return dateSession;}
    public Integer getMinutesStudied() {return minutesStudied;}
    public Integer getTasksCompleted() {return tasksCompleted;}
    public Integer getUserId() {return userId;}
}
